### PR TITLE
[SPARK-8457][ML]NGram Documentation

### PR DESCRIPTION
**Pull request by feynmanliang**
Add documentation for NGram feature transformer.

### CODE REVIEW

File: docs/ml-features.md [hunk 1]
```diff
@@ -288,6 +288,82 @@ for words_label in wordsDataFrame.select("words", "label").take(3):
 </div>
 
 
+## $n$-gram
+
+An [n-gram](https://en.wikipedia.org/wiki/N-gram) is a sequence of $n$ tokens (typically words) for some integer $n$. The [NGram](api/scala/index.html#org.apache.spark.ml.feature.NGram) class can be used to transform input features into $n$-grams.
```
Comment (jkbradley) [inline docs/ml-features.md]: I've started to put the link for the API docs under each language's codetab.  See StringIndexer's doc in this file for an example.

File: docs/ml-features.md [hunk 2]
```diff
@@ -288,6 +288,82 @@ for words_label in wordsDataFrame.select("words", "label").take(3):
...
+
+`NGram` takes as input a sequence of strings (e.g. the output of a [Tokenizer](api/scala/index.html#org.apache.spark.ml.feature.Tokenizer)).  The parameter `n` is used to determine the number of terms in each $n$-gram. The output will consist of a sequence of $n$-grams where each $n$-gram is represented by a space-delimited string of $n$ consecutive words.  If the input sequence contains less than `n` strings, no output is produced.
```
Comment (jkbradley) [inline docs/ml-features.md]: Could link to this page's section instead (once API doc links are moved to codetabs section).

"less" --> "fewer"

File: docs/ml-features.md [hunk 3]
```diff
@@ -288,6 +288,82 @@ for words_label in wordsDataFrame.select("words", "label").take(3):
...
+
+<div class="codetabs">
```
Comment (jkbradley) [inline docs/ml-features.md]: To confirm, have you run the Scala & Python examples in the shells, and tested the Java example in a test class?  If needed, you can add the Java test class as a unit test (modified not to print anything).

File: docs/ml-features.md [hunk 4]
```diff
@@ -288,6 +288,82 @@ for words_label in wordsDataFrame.select("words", "label").take(3):
...
+ngramDataFrame.select("ngrams", "label").take(3).foreach(println)
+{% endhighlight %}
+</div>
+
+<div data-lang="java" markdown="1">
+{% highlight java %}
+import com.google.common.collect.Lists;
+
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.ml.feature.Tokenizer;
```
Comment (jkbradley) [inline docs/ml-features.md]: don't need

need to import NGram

File: docs/ml-features.md [hunk 5]
```diff
@@ -288,6 +288,82 @@ for words_label in wordsDataFrame.select("words", "label").take(3):
...
+  new StructField("label", DataTypes.DoubleType, false, Metadata.empty()),
+  new StructField("words", DataTypes.createArrayType(DataTypes.StringType), false, Metadata.empty())
+});
+DataFrame wordDataFrame = sqlContext.createDataFrame(jrdd, schema);
+NGram ngramTransformer = new NGram().setInputCol("words").setOutputCol("ngrams");
+DataFrame ngramDataFrame = ngramTransformer.transform(wordDataFrame);
+for (Row r : ngramDataFrame.select("ngrams", "label").take(3)) {
+  java.util.List<String> ngrams = r.getList(0);
+  for (String ngram : ngrams) System.out.print(ngram + " --- ");
+    System.out.println();
```
Comment (jkbradley) [inline docs/ml-features.md]: fix indentation

File: docs/ml-features.md (hunk 3 above)
Comment (feynmanliang) [inline docs/ml-features.md]: Yes, tested them all. The Scala version returns Streams so the `println` looks weird but it is still correct.

File: docs/ml-features.md (hunk 2 above)
Comment (feynmanliang) [inline docs/ml-features.md]: What do you mean by link the page's section once API doc links moved?

File: docs/ml-features.md (hunk 4 above)
Comment (feynmanliang) [inline docs/ml-features.md]: OK.

File: docs/ml-features.md (hunk 5 above)
Comment (feynmanliang) [inline docs/ml-features.md]: OK.

File: docs/ml-features.md [hunk 6]
```diff
@@ -288,6 +288,88 @@ for words_label in wordsDataFrame.select("words", "label").take(3):
 </div>
 
 
+## $n$-gram
+
+An [n-gram](https://en.wikipedia.org/wiki/N-gram) is a sequence of $n$ tokens (typically words) for some integer $n$. The `NGram` class can be used to transform input features into $n$-grams.
+
+`NGram` takes as input a sequence of strings (e.g. the output of a [Tokenizer](api/scala/index.html#org.apache.spark.ml.feature.Tokenizer)).  The parameter `n` is used to determine the number of terms in each $n$-gram. The output will consist of a sequence of $n$-grams where each $n$-gram is represented by a space-delimited string of $n$ consecutive words.  If the input sequence contains fewer than `n` strings, no output is produced.
```
Comment (jkbradley) [inline docs/ml-features.md]: When I wrote
> Could link to this page's section instead (once API doc links are moved to codetabs section).

I meant that you could change the Tokenizer link to ```(ml-features.html#tokenizer)```.  (That was ambiguous.)

File: docs/ml-features.md [hunk 7]
```diff
@@ -288,6 +288,88 @@ for words_label in wordsDataFrame.select("words", "label").take(3):
...
+
+val wordDataFrame = sqlContext.createDataFrame(Seq(
+  (0, Array("Hi", "I", "heard", "about", "Spark")),
+  (1, Array("I", "wish", "Java", "could", "use", "case", "classes")),
+  (2, Array("Logistic", "regression", "models", "are", "neat"))
+)).toDF("label", "words")
+
+val ngram = new NGram().setInputCol("words").setOutputCol("ngrams")
+val ngramDataFrame = ngram.transform(wordDataFrame)
+ngramDataFrame.select("ngrams", "label").take(3).foreach(println)
```
Comment (jkbradley) [inline docs/ml-features.md]: If this looks weird, could you please modify it to be prettier?

**Comment by jkbradley** [comment]
Could you please put links for the API docs under the Java and Python codetabs?

File: docs/ml-features.md (hunk 6 above)
Comment (feynmanliang) [inline docs/ml-features.md]: OK

File: docs/ml-features.md (hunk 7 above)
Comment (feynmanliang) [inline docs/ml-features.md]: OK

**Comment by jkbradley** [comment]
LGTM pending tests

**Comment by jkbradley** [comment]
Merging with master
Thanks!

**Pull request closed without merging**